docs/datatypes/concurrency/index.md

```diff
@@ -1,13 +1,100 @@
+
+The _Shared Memory_ model has two main solutions:
+
+1. **Lock-Based** — In the locking model, the general primitives for synchronization are ـlocksـ, that control access to critical sections. When a thread wants to modify the critical section, it acquires the lock and says I'm the only thread that is allowed to modify the state right now, after it's work finished it unlocks the critical section and says I'm done, all other threads can modify this memory section.
+
+2. **Non-blocking** — Non-blocking algorithms usually use hardware-intrinsic atomic operations like `compare-and-swap` (CAS), without using any locks. This method follows an optimistic design with a transactional memory mechanism to roll back in conflict situations.
+
+## Implication of Locking Mechanism
+
+There are lots of drawback with lock-based concurrency:
+
+1. Incorrect use of locks can lead to deadlock. We need to care about the locking orders. If we don't place the locks in the right order, we may encounter a deadlock situation.
+
+2. Identifying the critical section of a code that is vulnerable to race conditions is overwhelming. We should always care about them and remember to lock everywhere it's required.
+
+3. It makes our software design very sophisticated to become scalable and reliable. It doesn't scale with program size and complexity.
+
+4. To prevent missing the releasing of the acquired locks, we should always care about exceptions and error handling inside locking sections. 
+
+5. The locking mechanism violates the encapsulation property of our pieces of programs. So systems that build with locking mechanism are difficult to compose without knowing about their internals.
+
+## Lock-free Concurrency Model
+
+As the lock-oriented programming does not compose and has lots of drawbacks, ZIO uses a _lock-free concurrency model_ which is a variation of non-blocking algorithms. The magic behind all of ZIO concurrency primitives is that they use CAS (_compare-and-set_) operation. 
+
+Let's see how the `modify` function of `Ref` is implemented without any locking mechanism:
+
+```scala mdoc:invisible
+import java.util.concurrent.atomic.AtomicReference
+import zio.UIO
+```
+
+```scala mdoc:silent
+  final case class Ref[A](value: AtomicReference[A]) { self =>
+    def modify[B](f: A => (B, A)): UIO[B] = UIO.effectTotal {
+      var loop = true
+      var b: B = null.asInstanceOf[B]
+      while (loop) {
+        val current = value.get
+        val tuple   = f(current)
+        b = tuple._1
+        loop = !value.compareAndSet(current, tuple._2)
+      }
+      b
+    }
+ }
+```
+
+The idea behind the `modify` is that a variable is only updated if it still has the same value as the time we had read the value from the original memory location. If the value has changed, it retries in the while loop until it succeeds. 
+
+## Advantage of Using ZIO Concurrency
```

Comment:
Another key point: ZIO concurrency structures, including those built with STM, are _safe for interruption_. This means that you can place them into timeouts, race them, and so forth, without leaking any resources.